src/models/tools/table/table-content.ts

```diff
@@ -171,14 +177,39 @@ export const TableContentModel = ToolContentModel
   .preProcessSnapshot(snapshot => {
     const s = snapshot as any;
     if (isTableImportSnapshot(s)) {
-      return { isImported: true, ...convertImportToSnapshot(s) };
+      return convertLegacyDataSet({ isImported: true, ...convertImportToSnapshot(s) });
     }
     if (s?.changes) {
-      return convertChangesToSnapshot(s.changes);
+      return convertLegacyDataSet(convertChangesToSnapshot(s.changes));
     }
-    return snapshot;
+    return convertLegacyDataSet(snapshot);
   })
+  .postProcessSnapshot(snapshot => {
+    const { importedDataSet, ...rest } = snapshot;
+    return { ...rest };
+  })
+  .views(self => ({
+    get sharedModel() {
+      const sharedModelManager = self.tileEnv?.sharedModelManager;
+      // Perhaps we should pass the type to getTileSharedModel, so it can return the right value
+      // just like findFirstSharedModelByType does
+      //
+      // For now we are checking the type ourselves, and we are assuming the shared model we want
+      // is the first one.
+      const firstSharedModel = sharedModelManager?.getTileSharedModels(self)?.[0];
+      if (!firstSharedModel || getType(firstSharedModel) !== SharedDataSet) {
+        return undefined;
+      }
+      return firstSharedModel as SharedDataSetType;
+    }
+  }))
   .views(self => ({
+    get dataSet() {
+      if (self.sharedModel) {
+        return self.sharedModel.dataSet;
+      }
+      return self.importedDataSet;
```

Comment:
I'd assume the importedDataSet will get converted to a sharedModel dataset, so wouldn't it better to return undefined until that happens?

Reply:
Previously, `dataSet` was defined like `dataSet: types.optional(DataSet, () => DataSet.create())`, so it was never undefined. There's a lot of code that assumes `TableContentModel.dataSet` will not be undefined, so this change results in a bunch of errors and would require changing a lot of other code. Is it worth changing now?

Reply:
If it seems like too much work then it's fine to skip it for now.